test/e2e/scale.go

```diff
@@ -0,0 +1,213 @@
+	Add(name string, start time.Time)
+}
+
+func ScaleToWithin(t *testing.T, logger *logging.BaseLogger, scale int, duration time.Duration, latencies Latencies) {
+	clients := Setup(t)
+
+	cleanupCh := make(chan test.ResourceNames, scale)
+	defer close(cleanupCh)
+
+	fopt := []ServiceOption{
+		// We set a small resource alloc so that we can pack more pods into the cluster.
+		func(svc *v1alpha1.Service) {
+			svc.Spec.RunLatest.Configuration.RevisionTemplate.Spec.Container.Resources = corev1.ResourceRequirements{
+				Limits: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("10m"),
+					corev1.ResourceMemory: resource.MustParse("50Mi"),
+				},
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("10m"),
+					corev1.ResourceMemory: resource.MustParse("20Mi"),
+				},
+			}
+		},
+		// See #2946 for why we do this.
+		func(svc *v1alpha1.Service) {
+			svc.Spec.RunLatest.Configuration.RevisionTemplate.ObjectMeta.Annotations = map[string]string{
+				"autoscaling.knative.dev/minScale": "1",
+				"autoscaling.knative.dev/maxScale": "1",
+			}
+		},
+	}
+	// These are the local (per-probe) and global (all probes) targets for the scale test.
+	// 95 = 19/20, so allow a single failure with the minimum number of probes, but expect
+	// us to have 3.5 9s overall.
+	const (
+		localSLO  = 0.95
+		globalSLO = 0.9995
+		minProbes = 20
+	)
+	pm := test.NewProberManager(logger, clients, minProbes)
+
+	timeoutCh := time.After(duration)
+
+	logger.Info("Creating new Services")
+	wg := pool.NewWithCapacity(50 /* maximum in-flight creates */, scale /* capacity */)
+	for i := 0; i < scale; i++ {
+		// https://golang.org/doc/faq#closures_and_goroutines
+		i := i
+
+		wg.Go(func() error {
+			names := test.ResourceNames{
+				Service: test.AppendRandomString(fmt.Sprintf("scale-%05d-%03d-", scale, i), logger),
+				Image:   "helloworld",
+			}
+
+			options := &test.Options{
+				// Give each request 10 seconds to respond.
+				// This is mostly to work around #2897
+				RevisionTimeoutSeconds: 10,
+				ReadinessProbe: &corev1.Probe{
+					Handler: corev1.Handler{
+						HTTPGet: &corev1.HTTPGetAction{
+							Path: "/",
+						},
+					},
+				},
+			}
+
+			// Start the clock for various waypoints towards Service readiness.
+			start := time.Now()
+			// Record the overall completion time regardless of success/failure.
+			defer latencies.Add("time-to-done", start)
+
+			svc, err := test.CreateLatestService(logger, clients, names, options, fopt...)
+			if err != nil {
+				t.Errorf("CreateLatestService() = %v", err)
+				return nil
+			}
+			// Record the time it took to create the service.
+			latencies.Add("time-to-create", start)
+			names.Route = serviceresourcenames.Route(svc)
+			names.Config = serviceresourcenames.Configuration(svc)
+
+			// Send it to our cleanup logic (below)
+			cleanupCh <- names
+
+			logger.Infof("Wait for %s to become ready.", names.Service)
+			var domain string
```

Comment:
I, personally, prefer `domain := ""`, but up to you :)

Reply:
I think I prefer this.